pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,662 @@
+		AddFunc: func(obj interface{}) {
+			c.enqueueLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.enqueueLease(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	leaseCandidateSynced, err := leaseCandidateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueCandidate(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseSynced = leaseSynced.HasSynced
+	c.leaseCandidateSynced = leaseCandidateSynced.HasSynced
+	return c, nil
+}
+
+func (c *Controller) enqueueLease(obj any) {
+	if lease, ok := obj.(*v1.Lease); ok {
+		key, err := controller.KeyFunc(lease)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+		}
+		c.leaseQueue.Add(key)
+	}
+}
+
+func (c *Controller) enqueueCandidate(obj any) {
+	if lease, ok := obj.(*v1alpha1.LeaseCandidate); ok {
+		_ = c.reconcileLeaseCandidate(lease)
+	}
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	klog.Infof("runWorker")
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) runElectionWorker(ctx context.Context) {
+	klog.Infof("runElectionWorker")
+	for c.processNextElectionItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextElectionItem(ctx context.Context) bool {
+	key, shutdown := c.electionQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.electionQueue.Done(key)
+
+	err := c.runElection(ctx, key)
+	utilruntime.HandleError(err)
+
+	return true
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := c.leaseQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.leaseQueue.Done(key)
+
+	err := func() error {
+		return c.reconcile(ctx, key)
+	}()
+
+	if err == nil {
+		c.leaseQueue.AddAfter(key, requeueInterval)
+		return true
+	}
+
+	utilruntime.HandleError(err)
+	klog.Infof("processNextWorkItem.AddRateLimited: %v", key)
+	c.leaseQueue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *Controller) reconcileLeaseCandidate(lc *v1alpha1.LeaseCandidate) error {
+	if lc == nil {
+		return nil
+	}
+	klog.Infof("reconcile found canLead label namespace=%q, name=%q: %q", lc.Namespace, lc.Name, lc.Spec.LeaseName)
+	leaderLeaseID := leaderLeaseID{lc.Namespace, lc.Spec.LeaseName}
+
+	// Ignore candidates that transitioned to Pending because reelection is already in progress
+	if lc.Spec.PingTime != nil {
+		return nil
+	}
+
+	// Updates during election will be marked as ack
+	status := c.getLS(leaderLeaseID)
+	if status.candidatesAck != nil {
+		status.candidateAck(lc)
+		return nil
+	}
+	// The current discriminating criteria is inElection vs not in election.
+	// Not all updates require election and more short circuiting logic would be beneficial.
+	c.electionQueue.Add(leaderLeaseID)
+	return nil
```

Comment:
Would make this much clearer, this code is crucial.

```suggestion
	status := c.getLS(leaderLeaseID)
	switch {
	case status.candidatesAck != nil:
		// ongoing leader election
		status.candidateAck(lc)
	default:
		// no election going on, start new one.
		// Not all updates require election and more short circuiting logic would be beneficial.
		c.electionQueue.Add(leaderLeaseID)
	}
	return nil
```

Reply:
What I really don't like is that a slow candidate that needs >5s to answer will cause an infinite loop of elections.

Reply:
I've optimized this to check the state before running an election. If this slow candidate is the most optimal candidate, then yes we will have that loop. Otherwise if the leader is already the most optimal, election is shortcircuited even if it is added to the electionQueue.